pkg/fleet/internal/oci/download.go

```diff
@@ -0,0 +1,236 @@
+	// the image URL is parsed as a digest to ensure we use the <repository>/<image>@<digest> format
+	ref, err := name.ParseReference(url, name.StrictValidation)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse ref: %w", err)
+	}
+	index, err := remote.Index(ref, remote.WithContext(ctx), remote.WithAuthFromKeychain(d.keychain), remote.WithTransport(httptrace.WrapRoundTripper(d.client.Transport)))
+	if err != nil {
+		return nil, fmt.Errorf("could not download image: %w", err)
+	}
+	return d.downloadIndex(index)
+}
+
+func (d *Downloader) downloadFile(path string) (oci.Image, error) {
+	layoutPath, err := layout.FromPath(path)
+	if err != nil {
+		return nil, fmt.Errorf("could not get layout from path: %w", err)
+	}
+	imageIndex, err := layoutPath.ImageIndex()
+	if err != nil {
+		return nil, fmt.Errorf("could not get image index: %w", err)
+	}
+	return d.downloadIndex(imageIndex)
+}
+
+func (d *Downloader) downloadIndex(index oci.ImageIndex) (oci.Image, error) {
+	platform := oci.Platform{
+		OS:           runtime.GOOS,
+		Architecture: runtime.GOARCH,
+	}
+	indexManifest, err := index.IndexManifest()
+	if err != nil {
+		return nil, fmt.Errorf("could not get index manifest: %w", err)
+	}
+	for _, manifest := range indexManifest.Manifests {
+		if manifest.Platform != nil && !manifest.Platform.Satisfies(platform) {
+			continue
+		}
+		image, err := index.Image(manifest.Digest)
+		if err != nil {
+			return nil, fmt.Errorf("could not get image: %w", err)
+		}
+		return image, nil
+	}
+	return nil, fmt.Errorf("no matching image found in the index")
+}
+
+// ExtractLayers extracts the layers of the downloaded package with the given media type to the given directory.
+func (d *DownloadedPackage) ExtractLayers(mediaType types.MediaType, dir string) error {
+	layers, err := d.Image.Layers()
+	if err != nil {
+		return fmt.Errorf("could not get image layers: %w", err)
+	}
+	for _, layer := range layers {
+		layerMediaType, err := layer.MediaType()
+		if err != nil {
+			return fmt.Errorf("could not get layer media type: %w", err)
+		}
+		if layerMediaType == mediaType {
+			uncompressedLayer, err := layer.Uncompressed()
+			if err != nil {
+				return fmt.Errorf("could not uncompress layer: %w", err)
+			}
+			err = tar.Extract(uncompressedLayer, dir, layerMaxSize)
+			if err != nil {
+				return fmt.Errorf("could not extract layer: %w", err)
+			}
+		}
+	}
+	return nil
+}
+
+// WriteOCILayout writes the image as an OCI layout to the given directory.
+func (d *DownloadedPackage) WriteOCILayout(dir string) error {
+	layoutPath, err := layout.Write(dir, empty.Index)
+	if err != nil {
+		return fmt.Errorf("could not write layout: %w", err)
+	}
+	err = layoutPath.AppendImage(d.Image)
+	if err != nil {
+		return fmt.Errorf("could not append image to layout: %w", err)
+	}
+	return nil
+}
+
+// PackageURL returns the package URL for the given site, package and version.
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 Thought:** ‏For prod we could theoretically use the registry in each site's cloud account